### ISSUE TITLE
dSYMs contain obfuscated symbols in the symbol table section when `--obfuscate` flag is on for iOS build

### ISSUE THREAD

**Issue by themiswang** [flutter/flutter#124715]
## Summary
When build iOS app with `--obfuscate --split-debug-info` flag on, `App.framework.dSYM` contains obfuscated symbols in the symbol table. For the symbolication tools such as breakpad, since it reads from DWARF so there is no problem. But there are other tools read from this symbol table which will results in returning obfuscated stack frame symbolication.

## Reproduction steps

- Build the app with `flutter build ios --release --obfuscate --split-debug-info=<FILE_PATH>` and `flutter build ios --release --split-debug-info=<FILE_PATH>` respectively. Get the `App.framework.dSYM` file for each build.

- If you compare their DWARFs there are no major differences between these two dSYMs.
 You can dump the DWARF information by doing:
  ```
  $ dwarfdump App.framework.dSYM/Contents/Resources/DWARF/App
  ```
  But if you checking the symbol table by doing:
  ```
  $ dsymutil --symtab App.framework.dSYM/Contents/Resources/DWARF/App
  ```
  The obfuscated dSYM contains obfuscated symbols which looks like this:
  <img width="727" alt="Screen Shot 2023-04-12 at 2 30 22 PM" src="https://user-images.githubusercontent.com/16548721/231551733-593d07f1-c6c8-470d-9e4e-6a42be0e3fd3.png">

## Expected
The symbol table should contains deobfuscated symbols even when `--obfuscate` flag on. We are also interested the intention or reason for the existing behaviour.

**Comment by darshankawar** [comment]
Thanks for the report @themiswang 
Can you take a look at [this](https://github.com/flutter/flutter/issues/116493) issue and underlying comments and see if it helps in your case further ?

**Comment by darshankawar** [comment]
Thanks for the feedback. I am going ahead and keep this issue open for team's insights on query raised above and actual and expected behavior reported.




> ## Expected
> The symbol table should contains deobfuscated symbols even when `--obfuscate` flag on. We are also interested the intention or reason for the existing behaviour.

**Comment by jmagman** [comment]
Apologies for the delay, we don't need an internal bug.

> The symbol table should contains deobfuscated symbols even when `--obfuscate` flag on.

This code was recently changed with https://github.com/flutter/flutter/pull/101586 and https://github.com/flutter/flutter/pull/111264.  I believe @mkustermann looked at the former.  

@christopherfujino @vashworth might be worth following up with the dart team about what the intended behavior of `--obfuscate` would be on the dSYM.  @jonahwilliams may know off the top of his head too.

cc @vaind

**Comment by vaind** [comment]
Hmm, that's interesting, I'm pretty I guess it worked fine when integrated into Sentry because (I assume), the symbolication there is using dwarf names. 

Also, I'm afraid I don't know what's the "right" approach (if there's any) for what should be in the `symtab`.

Additionally, if you look at how this is currently implemented, it's using `dsymutil` to extract dSYM from a compiled binary before stripping it - maybe there're flags to dsymutil to make it update the symbol table you see dumped from `dsymutil --symtab`?

https://github.com/flutter/flutter/blob/8ec5001ac2614e5dcc3eac8242896bfc7ea4e5a5/packages/flutter_tools/lib/src/base/build.dart#L315-L330

**Comment by themiswang** [comment]
Hey @vaind,

I did more research on this. When `--obfuscate` build option is on, before build the binary, I assume there is a process of obfuscate symbols, that's probably the reason why in the elf symbol table we see this all obfuscated symbols. For native iOS, how Swift deal with obfuscation is they will mangle the function name to some human unreadable format, but there is function provide to demangle the symbol. Wonder how obfuscation works on dart and does it has the similar strategy for de-obfuscate the symbols.

Attach a sample for native iOS how Swift symbol looks like:
<img width="1178" alt="Screenshot 2023-05-10 at 10 32 37 AM" src="https://github.com/flutter/flutter/assets/16548721/4e6bd630-41e7-4b76-9974-f8ad953b5a3d">

**Comment by mkustermann** [comment]
Apologies for late reply.

Enabling obfuscation in Dart will obfuscate class names and function names. This affects various things - not just stack traces. For example it also affects `<obj>.runtimeType.toString()`. That means even if a binary is stripped, there are symbols remaining in the stripped AOT snapshot that have to be obfuscated iff `--obfuscate` is enabled.

Due to this we obfuscate class/function names early on in our AOT compiler. Anything that comes later will use those obfuscated names. When emitting DWARF in the backend we explicitly map obfuscated symbols back to the original.
=> We could change the backend emitting the static symbol table to do the same.

For flutter specifically it seems for

* **iOS/macOS**: Flutter builds targeting Apple will tell the AOT compiler to emit the application as an assembly file. This assembly will already today contain unobfuscated DWARF (if `--split-debug-info` is used). Flutter tools will then compile the assembly to a shared library, extract symbols and DWARF to a `.dSYM` and strip the binary (by-default).

* **Android/Rest**: Flutter builds targeting non-Apple will tell the AOT compiler to emit the application as an ELF file. Flutter will then (by-default) tell our AOT compiler to strip the resulting ELF file, so we won't emit DWARF in it (we should also double-check we don't emit static symbols in this case).

Since stripping is enabled by default on both (for Apple platform the flutter tools does stripping, for non-Apple platform the AOT compiler emitting ELF does stripping) this will only cause an issue if someone explicitly uses `--obfuscate --extra-gen-snapshot-options=--no-strip` and then deploys such an app (which would have unobfuscated static symbols). That's maybe not a real concern as users intentionally leave symbol information in the resulting binary.

/cc @sstrickl Could you confirm the above is correct? Could we go ahead and make the static symbols unobfuscated (and ensure we omit static symbols if `--strip` is enabled - just as in dwarf)?

**Comment by sstrickl** [comment]
> /cc @sstrickl Could you confirm the above is correct? Could we go ahead and make the static symbols unobfuscated (and ensure we omit static symbols if `--strip` is enabled - just as in dwarf)?

Yes, I believe we should be able to have the static symbol generation code use unobfuscated strings, we'll just need to feed the appropriate information we already generate for the DWARF writer into the BlobImageWriters/AssemblyImageWriters so they can use that to deobfuscate symbol names before writing. Unless I'm forgetting something from the last time I touched that code, should be fairly straightforward.

**Comment by sstrickl** [comment]
Working on this now.

**Comment by mkustermann** [comment]
@sstrickl Consider verifying that `flutter build --release` doesn't leave any unstripped static symbols in the resulting ELF/MachO files? Then we can close this issue?

**Comment by themiswang** [comment]
Hey @sstrickl,

Also let me know if there is a need to update dart or flutter version in order to reflect the change. I'll need to working on other dependency change and announce with our customers as well. Thank you!

**Comment by mkustermann** [comment]
I believe this has been rolled all the way up to [flutter/flutter](https://github.com/flutter/flutter) now.

Closing this issue.

> Also let me know if there is a need to update dart or flutter version in order to reflect the change. 

@themiswang Could you verify this works as expected for you in flutter `master` channel?

**Comment by themiswang** [comment]
Hey @mkustermann,

Sorry I am not a member on Flutter team so I do need a stable version for testing which can be downloaded from [get start](https://docs.flutter.dev/get-started/install/macos). Let me know if you know how to do it or the folks who familiar with the process so I can schedule some time for verification.

**Comment by themiswang** [comment]
Never mind, find the instruction: https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#how-to-change-channels. Confirm it works now on main, thank you folks!
